In [108]:
import pandas as pd
import numpy as np
import re
from datetime import datetime

In [109]:
df_raw = pd.read_excel('C:/Users/admin/Desktop/YRES/Andi Dong - Summer Camp Popularity/Summer Camp Vancouver/Summer_Camp_Programs-_Vancouver.xlsx', sheet_name = 'Vancouver Camp Data')

In [110]:
df_fee_raw = pd.read_excel("C:/Users/admin/Desktop/YRES/Andi Dong - Summer Camp Popularity/Summer Camp Vancouver/vancouver_daycamps_fees_2026.xlsx")

In [111]:
df = df_raw.copy()

In [112]:
df_fee = df_fee_raw.copy()

In [113]:
df.head(10)

,Responsible,Category,Name,Age,Location,Date,Start Time,End Time,Cost (Resident),Cost (Non-Resident),...,Additional Notes,Status,Spots Left (Numeric),Date Cleaned,Activity ID,Openings (Listed),Status Marker (Raw),Weekdays,Num Days,Week (Date Start)
0,NaN,Dance/Performing Arts Summer Camp,1-Active Dance Sing/Jazz Funk/Hip Hop/KPOP (PM...,5-7 yrs,Dunbar Cmty Centre,27/07/2026 - 31/07/2026,12:00:00,15:00:00,NaN,NaN,...,Every day,Filling Fast,5,27/07/2026 - 31/07/2026,613400,5,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-07-27
1,NaN,Dance/Performing Arts Summer Camp,1-Active Dance Sing/Jazz Funk/Hip Hop/KPOP (PM...,5-7 yrs,Dunbar Cmty Centre,31/08/2026 - 04/09/2026,12:00:00,15:00:00,NaN,NaN,...,Every day,Open,10,31/08/2026 - 04/09/2026,621662,10,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-08-31
2,NaN,Dance/Performing Arts Summer Camp,"1-Active Tumble, Flex and Dance (AM Half Camp ...",5-13 yrs,Dunbar Cmty Centre,27/07/2026 - 31/07/2026,09:10:00,12:00:00,NaN,NaN,...,Every day,Filling Fast,4,27/07/2026 - 31/07/2026,613399,4,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-07-27
3,NaN,Dance/Performing Arts Summer Camp,"1-Active Tumble, Flex and Dance (AM Half Camp ...",5-13 yrs,Dunbar Cmty Centre,31/08/2026 - 04/09/2026,09:10:00,12:00:00,NaN,NaN,...,Every day,Open,7,31/08/2026 - 04/09/2026,621661,7,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-08-31
4,NaN,Dance/Performing Arts Summer Camp,"1-Active Tumble, Flex and Dance (AM Half Camp ...",8-14 yrs,Dunbar Cmty Centre,24/08/2026 - 28/08/2026,09:00:00,12:00:00,NaN,NaN,...,Every day,Open,10,24/08/2026 - 28/08/2026,613722,10,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-08-24
5,NaN,Dance/Performing Arts Summer Camp,"1-Active Tumble, Flex and Dance (AM Half Camp ...",8-14 yrs,Dunbar Cmty Centre,31/08/2026 - 04/09/2026,09:00:00,12:00:00,NaN,NaN,...,Every day,Open,10,31/08/2026 - 04/09/2026,621659,10,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-08-31
6,NaN,Dance/Performing Arts Summer Camp,"1-Active Tumble, Flex and Dance (PM Half Camp ...",5-7 yrs,Dunbar Cmty Centre,20/07/2026 - 24/07/2026,12:00:00,15:00:00,NaN,NaN,...,Every day,Open,9,20/07/2026 - 24/07/2026,613398,9,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-07-20
7,NaN,Dance/Performing Arts Summer Camp,"1-Active Tumble, Flex and Dance (PM Half Camp ...",8-14 yrs,Dunbar Cmty Centre,17/08/2026 - 21/08/2026,12:30:00,15:30:00,NaN,NaN,...,Every day,Open,9,17/08/2026 - 21/08/2026,613403,9,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-08-17
8,NaN,Dance/Performing Arts Summer Camp,Acrobatic Dance Camp,5-8 yrs,Roundhouse Cmty Arts and Rec Centre,04/08/2026 - 07/08/2026,14:30:00,15:30:00,NaN,NaN,...,"Every Tue, Wed, Thu, Fri",Open,9,04/08/2026 - 07/08/2026,610696,9,NaN,"Tue,Wed,Thu,Fri",4,2026-08-04
9,NaN,Dance/Performing Arts Summer Camp,"Act, Dance, Sing FUN! Camp",7-14 yrs,Kerrisdale Cmty Centre,06/07/2026 - 10/07/2026,12:30:00,15:30:00,255.0,NaN,...,Every day,Open,9,06/07/2026 - 10/07/2026,605503,9,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-07-06


In [114]:
df_fee.head(10)

,Activity ID,Cost
0,613400,$238.00
1,621662,$238.00
2,613399,$238.00
3,621661,$238.00
4,613722,$238.00
5,621659,$238.00
6,613398,$238.00
7,613403,$238.00
8,610696,$105.00
9,605503,Free


In [115]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1769 entries, 0 to 1768
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Responsible           0 non-null      float64
 1   Category              1769 non-null   object 
 2   Name                  1769 non-null   object 
 3   Age                   1769 non-null   object 
 4   Location              1741 non-null   object 
 5   Date                  1768 non-null   object 
 6   Start Time            1769 non-null   object 
 7   End Time              1769 non-null   object 
 8   Cost (Resident)       296 non-null    float64
 9   Cost (Non-Resident)   0 non-null      float64
 10  Descriptions          0 non-null      float64
 11  Spots left            1769 non-null   object 
 12  Additional Notes      1769 non-null   object 
 13  Status                1769 non-null   object 
 14  Spots Left (Numeric)  1769 non-null   int64  
 15  Date Cleaned         

In [116]:
df_fee.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1769 entries, 0 to 1768
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Activity ID  1769 non-null   int64 
 1   Cost         1769 non-null   object
dtypes: int64(1), object(1)
memory usage: 27.8+ KB


In [117]:
df.isna().all(axis=1).sum()

np.int64(0)

In [118]:
df.duplicated().sum()

np.int64(0)

In [119]:
df['Category'].unique() # 可以发现此处有十一个分类！太复杂了！让AI继续帮简化！

array(['Dance/Performing Arts Summer Camp', 'Other/Mixed Summer Camp',
       'Recreation/Day Camp', 'Outdoor/Adventure Summer Camp',
       'Creative Arts Summer Camp', 'Sports Summer Camp',
       'Technology and Engineering Summer Camp', 'Music Summer Camp',
       'Science Summer Camp', 'Performing Arts Summer Camp',
       'Education/Enrichment Summer Camp'], dtype=object)

In [120]:
import re

# Step 1: 删除 41 条 'New' (如果已经删过这步会 no-op)
before = len(df)
df = df[df['Name'] != 'New'].reset_index(drop=True)
print(f"删除 {before - len(df)} 条 'New'，剩余 {len(df)} 行")


def categorize(name):
    """
    Hybrid 分类逻辑：
    - 优先级 1 (最高): 基于 Chrome 手工核查得到的关键词 override
    - 优先级 2: 基于 keyword regex 的自动规则
    - 优先级 3 (兜底): General Day Camp / Other/Special
    """
    if not isinstance(name, str) or name.strip() == '':
        return 'Other/Special'
    n = name.lower()

    # ============================================================
    # 手工 override —— 基于实际查看 camp 描述得到的规则
    # 这些 camp 名字中没有明显类别关键词，但经核查后有专项归属
    # ============================================================

    # Sports override（飞盘/匹克球/体操/武术/篮球专项/体能/瑜伽等）
    if any(kw in n for kw in [
        'ultimate frisbee', 'pickleball', 'game ready crazy sports',
        'high 5 sports', 'rhythmic gymnastics', 'recreational gymnastics',
        'self defence', 'future bounce', 'shooters club',
        'kinetic kids', 'mindfulness & movement'
    ]):
        return 'Sports'

    # STEM override（各类机器人品牌/太空科学/工程设计）
    if any(kw in n for kw in [
        'mindstorms', 'wedo', 'spike robotics', 'robotics 1',
        "tomorrow's playground", 'steam 4 kids', 'space blast off',
        'space explorer', 'space explorers', 'inventor',
        'design and build', 'steamers'
    ]):
        return 'STEM'

    # Arts — Creative override（卡通/角色设计/电影制作/艺术/烹饪/创意设计）
    if any(kw in n for kw in [
        'cartooning', 'character design', 'young moviemakers',
        'little artist', 'famous artists remix', 'creative remix',
        'creative cooks', 'little cooks', 'fabric camp',
        'stories in the studio', 'amusement parks camp',
        'world cup cities', 'escape artist'
    ]):
        return 'Arts — Creative'

    # Arts — Performing/Dance override（主题舞蹈/音乐律动）
    if any(kw in n for kw in [
        'bluey', 'peppa', 'wickedly good time',
        'mini performance', 'zumbini'
    ]):
        return 'Arts — Performing/Dance'

    # Other/Special override（学术/识字/leadership/public speaking/empowerment）
    if any(kw in n for kw in [
        'math & literacy', 'math and literacy', 'ready to write',
        'ready for kindergarten', 'pre-kindergarten',
        'ready for the bell', 'empowerment camp',
        'hummingbird', 'kingfisher',
        'public speaking', 'entrepreneur mindset',
        'mindset masters', 'money masters', 'leadership'
    ]):
        return 'Other/Special'

    # ============================================================
    # 自动规则
    # ============================================================

    if re.search(r'\bsupported\b|1:1|\bpro-?d\b', n):
        return 'Other/Special'

    if (re.search(r'\b(robot|coding|code|programming|scratch|minecraft|roblox|'
                  r'python|ai\b|stem|engineer|engineering|computer|tech|'
                  r'game\s?design|3d|inventor|invention|science|lab|chemistry|'
                  r'biology|physics|forensic|chess|efk|architect|architecture)\b', n)
        or 'lego' in n or 'brick' in n or 'construction' in n or 'byte camp' in n):
        return 'STEM'

    if re.search(r'\b(dance|ballet|hip[\s-]?hop|jazz|kpop|k-pop|zumb|tumble|'
                 r'acrobat|choreograph|popping|locking|breaking|theatre|'
                 r'theater|drama|musical|act(?:ing)?|sing|singer|improv|'
                 r'broadway|play\s?making|music|guitar|piano|drum|band|'
                 r'ukulele|keyboard|dj\b|song)\b', n):
        return 'Arts — Performing/Dance'

    if re.search(r'\b(art\b|arts\b|craft|draw|drawing|paint|pottery|clay|'
                 r'sculpt|ceramic|collage|photograph|animation|anime|manga|'
                 r'film|filmmaking|media|fashion|sewing|cook|cooking|chef|'
                 r'baking|bake|culinary|kulinary|food|cartoon|cartoons|writing|'
                 r'storytelling|author|story|fairy|fairytale|superhero|video)\b', n):
        return 'Arts — Creative'

    if re.search(r'\b(sport|basketball|soccer|baseball|football|hockey|tennis|'
                 r'volleyball|badminton|golf|cricket|rugby|lacrosse|athletic|'
                 r'skateboard|cycling|bmx|track|karate|judo|taekwondo|martial|'
                 r'boxing|fencing|climbing|parkour|ninja|sportball|swim|aqua|'
                 r'pool|diving|yoga|fitness|capoeira)\b', n):
        return 'Sports'

    if re.search(r'\b(outdoor|adventure|hike|hiking|kayak|canoe|paddle|sail|'
                 r'fishing|bushcraft|wilderness|nature|forest|trail|eco\b|'
                 r'environmental|ranger|rangers)\b', n):
        return 'Outdoor & Adventure'

    if re.search(r'\b(day\s?camp|summer\s?camp|camp\b|funseekers|kids\s?camp|'
                 r'preschool|explorer|hangout|before\s?care|after\s?care|'
                 r'recreation|rec\b|daycamp|safari|smiles|daze|sunseekers|'
                 r'sunsplash|glow)\b', n):
        return 'General Day Camp'

    return 'Other/Special'


df['Category'] = df['Name'].apply(categorize)
print("\n=== 最终分类分布 ===")
print(df['Category'].value_counts())
print(f"\nTotal: {len(df)}")

删除 41 条 'New'，剩余 1728 行

=== 最终分类分布 ===
Category
General Day Camp           572
Sports                     452
STEM                       212
Arts — Performing/Dance    189
Arts — Creative            168
Other/Special              116
Outdoor & Adventure         19
Name: count, dtype: int64

Total: 1728


## 📋 Category 分类说明

以下是 Vancouver Summer Camp 数据集重新分类后的 7 大类定义：

**📌 General Day Camp** — 综合型日托 camp，社区中心主流产品。每周主题不同，混合多种活动（手工 + 游戏 + 户外 + 运动），主要面向 6-12 岁儿童。包含各类 Day Camp / Summer Camp / Funseekers / Explorers / Safari / Preschool camp / Before-After Care 等。

**🏀 Sports** — 体育专项 camp。篮球、足球、网球、棒球、冰球、排球、羽毛球、高尔夫、武术（空手道/柔道/跆拳道/卡波耶拉）、攀岩、跑酷、游泳/水上运动、瑜伽/健身、Sportball 等。

**🔬 STEM** — 科学/技术/工程/数学类。LEGO 搭建、机器人、编程（Scratch/Python/Minecraft/Roblox）、游戏设计、3D 建模、AI、科学实验（化学/生物/物理/法医）、工程/建筑、国际象棋、Byte Camp（编程品牌）等。

**💃 Arts — Performing/Dance** — 表演艺术和舞蹈类。各类舞蹈（芭蕾/爵士/Hip-Hop/KPOP/街舞/Zumba/Tumbling/编舞）、戏剧/音乐剧、即兴剧、音乐教学（吉他/钢琴/鼓/乐队/声乐）等。

**🎨 Arts — Creative** — 创意和视觉艺术类。绘画、陶艺/雕塑、摄影、动画/动漫、电影/视频制作、时装/缝纫、烹饪/烘焙、卡通创作、创意写作/故事创作、童话改编等。

**🌲 Outdoor & Adventure** — 户外探险专项 camp。Discoveries Adventure、Preteen Adventure、Renfrew Rangers Outdoor Leadership 等名字里明确强调户外属性的项目。注：很多 General Day Camp 里也穿插户外活动，但只有**名字里明确户外定位**的才归入此类。

**♿ Other/Special** — 特殊需求 + 特殊形式 + 生活技能类 camp。
- 特殊需求：Supported Day Camp (1:1 辅助员陪伴)、Falaise Fun Finders、Supported Sunsplash 系列
- 特殊形式：Pro-D Day Camp (Professional Development 日托)
- 学术预备：Math & Literacy、Ready to Write、Ready for Kindergarten、Pre-Kindergarten、Ready for the Bell
- 领导力和公共演讲：Summer Leadership Camp、Youth Leadership、Public Speaking、TED-Style Speaking
- 企业家思维：Future Ready Minds 系列（Entrepreneur Mindset、Money Masters、Mindset Masters）
- 赋权项目：Fire & Flower Empowerment (Hummingbird / Kingfisher)，面向 LGBTQ2SIAP+ 青少女的免费申请制项目

In [121]:
df['Age']

0           5-7 yrs
1           5-7 yrs
2          5-13 yrs
3          5-13 yrs
4          8-14 yrs
           ...     
1723      11-14 yrs
1724    Under 5 yrs
1725    Under 5 yrs
1726    Under 5 yrs
1727    Under 5 yrs
Name: Age, Length: 1728, dtype: object

In [122]:
df['Age'].unique()

array(['5-7 yrs', '5-13 yrs', '8-14 yrs', '5-8 yrs', '7-14 yrs',
       '5-14 yrs', '6-13 yrs', '6-14 yrs', '5-15 yrs', '6-8 yrs',
       '9-13 yrs', '5-12 yrs', '6-12 yrs', '6-11 yrs', '7-12 yrs',
       '4-5 yrs', '8-12 yrs', '4-6 yrs', '3-5 yrs', '9-12 yrs',
       '5-11 yrs', '5-10 yrs', '6-7 yrs', '10-14 yrs', '11-14 yrs',
       '3-4 yrs', '6-9 yrs', '7-10 yrs', '9-14 yrs', '5-6 yrs', '3-6 yrs',
       '10-13 yrs', '6-17 yrs', '6-10 yrs', '8-9 yrs', '10-12 yrs',
       '11-15 yrs', '12-14 yrs', '12-15 yrs', '8-13 yrs', '8-11 yrs',
       '4-7 yrs', '13-16 yrs', '7-9 yrs', '8-10 yrs', '7-11 yrs',
       '7-13 yrs', '6-16 yrs', 'Under 4 yrs', 'Under 5 yrs',
       'Under 1 yrs', '1-3 yrs', '5-5 yrs', '10-15 yrs', '11-16 yrs',
       '5-9 yrs', '6-15 yrs', '11-13 yrs', '10-16 yrs', '11-17 yrs',
       '9-15 yrs'], dtype=object)

In [123]:
import re

# Step 1: 从 Age 字符串提取 min_age
def parse_min_age(age_str):
    if not isinstance(age_str, str):
        return None
    s = age_str.lower().strip()
    # "Under X yrs" → min = 0
    if s.startswith('under'):
        return 0
    # "X-Y yrs" 或 "X yrs" → 取前面的数字
    m = re.match(r'(\d+)', s)
    if m:
        return int(m.group(1))
    return None

df['Min Age'] = df['Age'].apply(parse_min_age)

# Step 2: 按 Min Age 分桶（3 档）
def assign_age_group(min_age):
    if pd.isna(min_age):
        return 'Unknown'
    if min_age <= 5:
        return 'Pre-K (0-5)'
    if min_age <= 9:
        return 'Child (6-9)'
    return 'Pre-teen/Teen (10+)'

df['Age Group'] = df['Min Age'].apply(assign_age_group)

# 查看结果
print("=== Age Group 分布（3 档，按 Min Age）===")
print(df['Age Group'].value_counts())
print(f"\nTotal: {len(df)}")
print(f"Min Age 缺失: {df['Min Age'].isna().sum()}")

=== Age Group 分布（3 档，按 Min Age）===
Age Group
Child (6-9)            996
Pre-K (0-5)            541
Pre-teen/Teen (10+)    191
Name: count, dtype: int64

Total: 1728
Min Age 缺失: 0


## 👶 Age Group 分类说明（按 Min Age 分桶）

本项目采用 **Min Age（最低参加年龄）** 作为年龄分桶依据。
原因：家长选 camp 的核心决策是"我家孩子够不够格参加"——min age 是准入门槛，
直接驱动报名决策，比 max age 或年龄跨度更有业务信号。

**Pre-K (0-5)** — `min_age ≤ 5`  
学前儿童 camp。包括 0-4 岁的亲子音乐课（Music with Marnie, Zumbini）、
Pre-Kindergarten 学前衔接 camp、Preschool Summer Camp 等。
对应加拿大 Kindergarten 入学前阶段。

**Child (6-9)** — `6 ≤ min_age ≤ 9`  
小学低年级主力客群。对应加拿大 Elementary Junior（K-3）阶段。
这是暑期 camp 市场的最大客群，家长暑期托管需求最旺盛。

**Pre-teen/Teen (10+)** — `min_age ≥ 10`  
小学高年级 + 初中 camp。对应加拿大 Elementary Senior (4-7) 到
Secondary 早期阶段。这个群体通常更独立，camp 内容也更专项化
（高阶运动训练、编程、Leadership 等）。

In [124]:
# 现在来看date列！
df['Date']

0       27/07/2026 - 31/07/2026
1       31/08/2026 - 04/09/2026
2       27/07/2026 - 31/07/2026
3       31/08/2026 - 04/09/2026
4       24/08/2026 - 28/08/2026
                 ...           
1723    17/08/2026 - 21/08/2026
1724    29/06/2026 - 03/07/2026
1725    04/08/2026 - 07/08/2026
1726    29/06/2026 - 03/07/2026
1727    24/08/2026 - 27/08/2026
Name: Date, Length: 1728, dtype: object

In [125]:
df[['Start Date', 'End Date']] = df['Date'].str.split(' - ', expand=True)

df['Start Date'] = pd.to_datetime(df['Start Date'], dayfirst=True)
df['End Date'] = pd.to_datetime(df['End Date'], dayfirst=True)

In [126]:
# 把费用dataset与主表合并
# 先检查key！
df_fee.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1769 entries, 0 to 1768
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Activity ID  1769 non-null   int64 
 1   Cost         1769 non-null   object
dtypes: int64(1), object(1)
memory usage: 27.8+ KB


In [127]:
df_fee['Activity ID'].isna().sum()

np.int64(0)

In [128]:
df_fee['Activity ID'].duplicated().sum()

np.int64(0)

In [129]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1728 entries, 0 to 1727
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Responsible           0 non-null      float64       
 1   Category              1728 non-null   object        
 2   Name                  1728 non-null   object        
 3   Age                   1728 non-null   object        
 4   Location              1700 non-null   object        
 5   Date                  1727 non-null   object        
 6   Start Time            1728 non-null   object        
 7   End Time              1728 non-null   object        
 8   Cost (Resident)       276 non-null    float64       
 9   Cost (Non-Resident)   0 non-null      float64       
 10  Descriptions          0 non-null      float64       
 11  Spots left            1728 non-null   object        
 12  Additional Notes      1728 non-null   object        
 13  Status            

In [130]:
df['Activity ID'].isna().sum()

np.int64(0)

In [131]:
df['Activity ID'].duplicated().sum()

np.int64(0)

In [132]:
# key列没有问题，可以合并！
df_final = pd.merge(df, df_fee, on = 'Activity ID', how = 'left')

In [133]:
df_final.head()

,Responsible,Category,Name,Age,Location,Date,Start Time,End Time,Cost (Resident),Cost (Non-Resident),...,Openings (Listed),Status Marker (Raw),Weekdays,Num Days,Week (Date Start),Min Age,Age Group,Start Date,End Date,Cost
0,NaN,Arts — Performing/Dance,1-Active Dance Sing/Jazz Funk/Hip Hop/KPOP (PM...,5-7 yrs,Dunbar Cmty Centre,27/07/2026 - 31/07/2026,12:00:00,15:00:00,NaN,NaN,...,5,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-07-27,5,Pre-K (0-5),2026-07-27,2026-07-31,$238.00
1,NaN,Arts — Performing/Dance,1-Active Dance Sing/Jazz Funk/Hip Hop/KPOP (PM...,5-7 yrs,Dunbar Cmty Centre,31/08/2026 - 04/09/2026,12:00:00,15:00:00,NaN,NaN,...,10,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-08-31,5,Pre-K (0-5),2026-08-31,2026-09-04,$238.00
2,NaN,Arts — Performing/Dance,"1-Active Tumble, Flex and Dance (AM Half Camp ...",5-13 yrs,Dunbar Cmty Centre,27/07/2026 - 31/07/2026,09:10:00,12:00:00,NaN,NaN,...,4,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-07-27,5,Pre-K (0-5),2026-07-27,2026-07-31,$238.00
3,NaN,Arts — Performing/Dance,"1-Active Tumble, Flex and Dance (AM Half Camp ...",5-13 yrs,Dunbar Cmty Centre,31/08/2026 - 04/09/2026,09:10:00,12:00:00,NaN,NaN,...,7,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-08-31,5,Pre-K (0-5),2026-08-31,2026-09-04,$238.00
4,NaN,Arts — Performing/Dance,"1-Active Tumble, Flex and Dance (AM Half Camp ...",8-14 yrs,Dunbar Cmty Centre,24/08/2026 - 28/08/2026,09:00:00,12:00:00,NaN,NaN,...,10,NaN,"Mon,Tue,Wed,Thu,Fri",5,2026-08-24,8,Child (6-9),2026-08-24,2026-08-28,$238.00


In [134]:
# 对location列做格式清洗+统一！
df_final['Location'] = df_final['Location'].str.strip().str.title()

In [135]:
df_final['Cost']

0       $238.00
1       $238.00
2       $238.00
3       $238.00
4       $238.00
         ...   
1723    $145.00
1724       Free
1725     $80.00
1726     $80.00
1727     $80.00
Name: Cost, Length: 1728, dtype: object

In [136]:
# 对cost列做清洗！
df_final['Cost'] = df_final['Cost'].replace('Free', '0')

In [137]:
# 2. 使用正则提取第一个数字（包括小数点）
# 这个正则会匹配类似 325.00 的部分，忽略后面括号里的内容
df_final['Cost'] = df_final['Cost'].str.extract(r'(\d+\.?\d*)')[0]

In [138]:
df_final['Cost'] = df_final['Cost'].astype(float)

In [139]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1728 entries, 0 to 1727
Data columns (total 27 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Responsible           0 non-null      float64       
 1   Category              1728 non-null   object        
 2   Name                  1728 non-null   object        
 3   Age                   1728 non-null   object        
 4   Location              1700 non-null   object        
 5   Date                  1727 non-null   object        
 6   Start Time            1728 non-null   object        
 7   End Time              1728 non-null   object        
 8   Cost (Resident)       276 non-null    float64       
 9   Cost (Non-Resident)   0 non-null      float64       
 10  Descriptions          0 non-null      float64       
 11  Spots left            1728 non-null   object        
 12  Additional Notes      1728 non-null   object        
 13  Status            

In [140]:
df_final[['Start Time', 'End Time']]

,Start Time,End Time
0,12:00:00,15:00:00
1,12:00:00,15:00:00
2,09:10:00,12:00:00
3,09:10:00,12:00:00
4,09:00:00,12:00:00
...,...,...
1723,09:00:00,15:30:00
1724,10:30:00,11:15:00
1725,09:45:00,10:30:00
1726,13:00:00,13:45:00


In [141]:
# 对这两列时间做清洗！
#df_final['Start Time'] = pd.to_timedelta(df_final['Start Time'], errors = 'coerce')
# df_final['End Time'] = pd.to_timedelta(df_final['End Time'], errors = 'coerce')
# 通过这一步操作反向发现了1567行异常数据集！所以rerun kernel来反过来检查并修正！

In [142]:
df_final.loc[[1567]]['End Time']

1567    2:30 AM on the next day
Name: End Time, dtype: object

In [143]:
# 删除第 1567 行，并重新排列索引
df_final = df_final.drop(1567).reset_index(drop=True)

In [144]:
# 对这两列时间做清洗！
df_final['Start Time'] = pd.to_timedelta(df_final['Start Time'])
df_final['End Time'] = pd.to_timedelta(df_final['End Time'])

In [145]:
df_final[['Start Time', 'End Time']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1727 entries, 0 to 1726
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype          
---  ------      --------------  -----          
 0   Start Time  1727 non-null   timedelta64[ns]
 1   End Time    1727 non-null   timedelta64[ns]
dtypes: timedelta64[ns](2)
memory usage: 27.1 KB


In [146]:
df_final['Status'].unique()

array(['Filling Fast', 'Open', 'Cancelled', 'Full', 'Almost Full'],
      dtype=object)

In [147]:
df_final['Spots Left (Numeric)'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 1727 entries, 0 to 1726
Series name: Spots Left (Numeric)
Non-Null Count  Dtype
--------------  -----
1727 non-null   int64
dtypes: int64(1)
memory usage: 13.6 KB


In [148]:
df_final['Spots Left (Numeric)'].isna().sum()

np.int64(0)

In [149]:
# 清洗spots left numeric行！
# 将 Status 为 'Cancelled' 的行对应的 'Spots Left (Numeric)' 修改为 -1
df_final.loc[df_final['Status'] == 'Cancelled', 'Spots Left (Numeric)'] = -1

In [150]:
df_final['Num Days'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 1727 entries, 0 to 1726
Series name: Num Days
Non-Null Count  Dtype
--------------  -----
1727 non-null   int64
dtypes: int64(1)
memory usage: 13.6 KB


In [151]:
df_final['Num Days'].unique()

array([5, 4, 3, 2, 1])

In [152]:
df_final['Week (Date Start)'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 1727 entries, 0 to 1726
Series name: Week (Date Start)
Non-Null Count  Dtype 
--------------  ----- 
1726 non-null   object
dtypes: object(1)
memory usage: 13.6+ KB


In [153]:
df_final['Week (Date Start)'].head()

0    2026-07-27
1    2026-08-31
2    2026-07-27
3    2026-08-31
4    2026-08-24
Name: Week (Date Start), dtype: object

In [154]:
# 将字符串转换为标准日期格式 (YYYY-MM-DD)
df_final['Week (Date Start)'] = pd.to_datetime(df_final['Week (Date Start)'])

In [155]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1727 entries, 0 to 1726
Data columns (total 27 columns):
 #   Column                Non-Null Count  Dtype          
---  ------                --------------  -----          
 0   Responsible           0 non-null      float64        
 1   Category              1727 non-null   object         
 2   Name                  1727 non-null   object         
 3   Age                   1727 non-null   object         
 4   Location              1699 non-null   object         
 5   Date                  1726 non-null   object         
 6   Start Time            1727 non-null   timedelta64[ns]
 7   End Time              1727 non-null   timedelta64[ns]
 8   Cost (Resident)       276 non-null    float64        
 9   Cost (Non-Resident)   0 non-null      float64        
 10  Descriptions          0 non-null      float64        
 11  Spots left            1727 non-null   object         
 12  Additional Notes      1727 non-null   object         
 13  Sta

In [156]:
# 用 End Time 减去 Start Time，并将结果转换为小时（小数格式）
df_final['Duration_Hours'] = (df_final['End Time'] - df_final['Start Time']).dt.total_seconds() / 3600

In [157]:
df_final['Duration_Hours'].describe()

count    1727.000000
mean        4.091575
std         2.344240
min         0.500000
25%         2.000000
50%         3.000000
75%         6.500000
max        10.500000
Name: Duration_Hours, dtype: float64

In [158]:
# 定义函数：根据时长划分类型
def classify_camp(hour):
    if hour >= 5:
        return 'Full Day'
    else:
        return 'Half Day'

# 创建新列 'Camp Type'
df_final['Camp Type'] = df_final['Duration_Hours'].apply(classify_camp)

# 看看分布情况
print(df_final['Camp Type'].value_counts())

Camp Type
Half Day    962
Full Day    765
Name: count, dtype: int64


In [ ]:
df_final.info()

In [ ]:
df_final.to_csv("YRES_Summer Camp_Vancouver_cleaned_data.csv", index=False, encoding="utf-8-sig")




## 📊 Data Quality Summary (数据质量报告)

### 1. 样本概况 (Data Volume)
* **总原始记录数**: 1769 条
* **最终分析记录数**: 1727 条 **删除了41条无法被分类category的类别**
* **排除记录数**: 1 条 (Index 1567: 录入格式异常，包含 "on the next day" 导致无法计算)

### 2. 字段缺失率分析 (Missing Values)
| 字段名称 | 缺失率 | 说明 |
| :--- | :--- | :--- |
| **Location** | 1.6% | 极少数项目未标注地点。 |
| **Start/End Time** | 0% | 时间字段已全部补全或清理，覆盖率 100%。 |
| **Spots Left** | 0% | 核心指标完整，可用于活跃度分析。 |

### 3. 已知数据局限性 (Data Limitations)
* **名额代理变量 (Proxy)**: 数据集没有提供原始总容量（Total Capacity），目前使用 `Spots Left (Numeric)` 作为报名的代理变量。
* **取消课程干扰**: 部分状态为 `Cancelled` 的课程名额显示异常，已统一处理为 `-1` 以便在统计中剔除。

### 4. 核心清洗决策 (Cleaning Decisions)
* **类型转换**: 将 `Cost` 列中的 "Free" 归零，并使用正则提取主价格，剔除了括号内的额外杂费以保持一致性。
* **时长计算**: 统一使用 `timedelta` 计算 `Duration_Hours`，并将结果划分为 **Half Day (<5h)** 和 **Full Day (>=5h)**。
* **异常处理**: 针对 `Status` 为 `Cancelled` 的项目，强制将其剩余名额设为 `-1`，避免在分析“最火热项目”时误入分母。
* **日期标准化**: 将 `Week (Date Start)` 转换为标准 `datetime` 格式，支持按周进行趋势分析。

---

### ✅ 导出指令 (Final Cell)
完成上述总结后，运行最后一行代码导出最终文件：

```python
# 最终确认并导出
df_final.to_csv("YRES_Summer_Camp_Vancouver_cleaned_data.csv", index=False, encoding="utf-8-sig")
print("✅ 数据清洗成功，文件已保存！")
```